File: lib/rgfa/line/common/field_data.rb
module RGFA::Line::Common::FieldData

  # @return [Array<Symbol>] name of the positional fields
  # @note these names are not always the field names
  #   in the specification,
  #   as these may be implemented as aliases to cope with
  #   different names for the same content in GFA1 vs GFA2
  # @api private
  def positional_fieldnames
    if @version.nil?
      raise RGFA::VersionError, "Version is not set"
    end
    self.class::POSFIELDS
  end

  # @return [Array<Symbol>] name of the defined tags
  def tagnames
    (@data.keys - positional_fieldnames)
  end

  # Set the value of a field.
  #
  # If a datatype for a new custom tag is not set,
  # the default for the value assigned to the field will be used
  # (e.g. J for Hashes, i for Integer, etc).
  #
  # @param fieldname [Symbol] the name of the field to set
  #   (positional field, predefined tag (uppercase) or custom tag (lowercase))
  # @raise [RGFA::FormatError] if +fieldname+ is not a
  #   valid predefined or custom tag name (and +validate[:tags]+)
  # @return [Object] +value+
  def set(fieldname, value)
    if @data.has_key?(fieldname) or predefined_tag?(fieldname)
      return set_existing_field(fieldname, value)
    elsif self.class::FIELD_ALIAS.has_key?(fieldname)
      return set(self.class::FIELD_ALIAS[fieldname], value)
    elsif (@validate == 0) or valid_custom_tagname?(fieldname)
      define_field_methods(fieldname)
      if !@datatype[fieldname].nil?
        return set_existing_field(fieldname, value)
      elsif !value.nil?
        @datatype[fieldname] = value.default_gfa_tag_datatype
        return @data[fieldname] = value
      end
    else
      raise RGFA::FormatError,
        "#{fieldname} is not an existing or predefined field or a "+
        "valid custom tag"
    end
  end

  # Get the value of a field
  # @param fieldname [Symbol] name of the field
  # @return [Object,nil] value of the field
  #   or +nil+ if field is not defined
  def get(fieldname)
    v = @data[fieldname]
    if v.kind_of?(String)
      t = field_datatype(fieldname)
      if t != :Z and t != :seq
        # value was not parsed or was set to a string by the user
        return (@data[fieldname] = v.parse_gfa_field(t, safe: @validate >= 2))
      else
         v.validate_gfa_field!(t, fieldname) if (@validate >= 5)
      end
    elsif !v.nil?
      if (@validate >= 5)
        t = field_datatype(fieldname)
        v.validate_gfa_field!(t, fieldname)
      end
    else
      dealiased_fieldname = self.class::FIELD_ALIAS[fieldname]
      return get(dealiased_fieldname) if !dealiased_fieldname.nil?
    end
    return v
  end

  # Value of a field, raising an exception if it is not defined
  # @param fieldname [Symbol] name of the field
  # @raise [RGFA::NotFoundError] if field is not defined
  # @return [Object,nil] value of the field
  def get!(fieldname)
    v = get(fieldname)
    raise RGFA::NotFoundError,
      "No value defined for tag #{fieldname}" if v.nil?
    return v
  end

  # Remove a tag from the line, if it exists; do nothing if it does not
  # @param tagname [Symbol] the tag name of the tag to remove
  # @return [Object, nil] the deleted value or nil, if the field was not defined
  def delete(tagname)
    if tagnames.include?(tagname)
      @datatype.delete(tagname)
      return @data.delete(tagname)
    else
      return nil
    end
  end

  private

  def set_existing_field(fieldname, value)
    if value.nil?
      @data.delete(fieldname)
    else
      if @validate >= 5
        field_or_default_datatype(fieldname, value)
        value.validate_gfa_field!(field_datatype(fieldname), fieldname)
      end
      @data[fieldname] = value
    end
  end

end

File: gfapy/line/common/field_data.py
import gfapy

class FieldData:

  @property
  def positional_fieldnames(self):
    """Name of the positional fields.
    Returns:
      str list

    Note:
      The field names can vary from the exact name of the
      fields in the specification, due to compatibility issues.
    """
    return self.__class__.POSFIELDS

  @property
  def tagnames(self):
    """Names of the tags defined in the line.
    Returns:
      str list
    """
    return [ x for x in self._data.keys() \
        if x not in self.positional_fieldnames ]

  @property
  def record_type(self):
    """The record type.

    The content of the first field of the line. In all predefined record types
    this is a single-letter upper case string.

    Returns:
      str
    """
    return self.__class__.RECORD_TYPE

  def set(self, fieldname, value):
    """Set the value of a field.

    If a datatype for a new custom tag is not set,
    the default for the value assigned to the field will be used
    (e.g. J for Hashes, i for Integer, etc).

    Parameters
    ----------
    fieldname : str
      The name of the field to set.
      (positional field, predefined or custom tag)

    Raises
    ------
    gfapy.FormatError
      If **fieldname** is not a valid predefined or
      custom tag name (and **validate["tags"]**).

    Returns
    -------
    object
      **value**
    """
    if fieldname in self._data or self._is_predefined_tag(fieldname):
      return self._set_existing_field(fieldname, value)
    elif fieldname in self.__class__.FIELD_ALIAS:
      return self.set(self.__class__.FIELD_ALIAS[fieldname], value)
    elif self.virtual:
      raise gfapy.RuntimeError("Virtual lines do not have tags")
    elif (self.vlevel == 0) or self._is_valid_custom_tagname(fieldname):
      self._define_field_methods(fieldname)
      if self._datatype.get(fieldname, None) is not None:
        return self._set_existing_field(fieldname, value)
      elif value is not None:
        self._datatype[fieldname] = \
            gfapy.Field._get_default_gfa_tag_datatype(value)
        self._data[fieldname] = value
        return self._data[fieldname]
    else:
      raise gfapy.FormatError(
        "{} is not a positional field,".format(fieldname)+
        "an existing tag, an alias, a predefined tag or a valid custom tag\n"+
        "positional fields: {}\n".format(", ".join(self.positional_fieldnames))+
        "existing tags: {}\n".format(", ".join(self.tagnames))+
        "aliases: {}\n".format(", ".join(self.__class__.FIELD_ALIAS.keys()))+
        "predefined tags: {}\n".format(", ".join(self.__class__.PREDEFINED_TAGS)))

  def get(self, fieldname):
    """
    Get the value of a field.

    Parameters
    ----------
    fieldname : str
      Name of the field.

    Returns
    -------
    object or None
      Value of the field or **None** if field is not defined.
    """
    v = self._data.get(fieldname, None)
    if isinstance(v, str):
      t = self._field_datatype(fieldname)
      if t != "Z" and t != "seq":
        # value was not parsed or was set to a string by the user
        self._data[fieldname] = gfapy.Field._parse_gfa_field(v, t,
                                                    safe = (self.vlevel >= 1),
                                                    fieldname = fieldname,
                                                    line = self)
        return self._data[fieldname]
      else:
        if (self.vlevel >= 3):
          gfapy.Field._validate_gfa_field(v, t, fieldname)
    elif v is not None:
      if (self.vlevel >= 3):
        t = self._field_datatype(fieldname)
        gfapy.Field._validate_gfa_field(v, t, fieldname)
    else:
      dealiased_fieldname = self.__class__.FIELD_ALIAS.get(fieldname, None)
      if dealiased_fieldname is not None:
        return self.get(dealiased_fieldname)
    return v

  def try_get(self, fieldname):
    """
    Value of a field, raising an exception if it is not defined.

    Parameters
    ----------
    fieldname : str
      Name of the field.

    Raises
    ------
    gfapy.NotFoundError
      If field is not defined.

    Returns
    -------
    object or None
      Value of the field.
    """
    v = self.get(fieldname)
    if v is None:
      raise gfapy.NotFoundError(
        "No value defined for tag {}".format(fieldname))
    return v

  def delete(self, tagname):
    """
    Remove a tag from the line, if it exists; do nothing if it does not.

    Parameters
    ----------
    tagname : Symbol
      The tag name of the tag to remove.

    Returns
    -------
    object or None
      The deleted value or None, if the field was not defined.
    """
    if tagname in self.tagnames:
      if tagname in self._datatype:
        self._datatype.pop(tagname)
      return self._data.pop(tagname)
    else:
      return None

  def _set_existing_field(self, fieldname, value, set_reference = False):
    renaming_connected = False
    if self._gfa:
      if not set_reference and \
        (fieldname in self.__class__.REFERENCE_FIELDS or \
         fieldname in self.__class__.BACKREFERENCE_RELATED_FIELDS):
        raise gfapy.RuntimeError(
          "The value of field '{}' cannot be changed, ".format(fieldname)+
          "as the line belongs to a GFA instance")
      if (fieldname == self.__class__.STORAGE_KEY) or \
        (self.__class__.STORAGE_KEY == "name" and \
        fieldname == self.__class__.NAME_FIELD):
         renaming_connected = True
         self._gfa._unregister_line(self)
    if value is None:
      if fieldname in self._data:
        self._data.pop(fieldname)
    else:
      if self.vlevel >= 3:
        self._field_or_default_datatype(fieldname, value)
        gfapy.Field._validate_gfa_field(value, self._field_datatype(fieldname),
            fieldname)
      self._data[fieldname] = value
    if renaming_connected:
      self._gfa._register_line(self)

  def _dealias_fieldname(self, fieldname):
    return self.__class__.FIELD_ALIAS.get(fieldname, fieldname)

  def _dealias_fieldnames(self, fieldnames):
    fieldnames[:] = map(self._dealias_fieldname, fieldnames)
